QUICK-FIX Update the PR review guide

plamut opened the pull request:
**NOTE:** There are still two pieces missing - the section on Docker and on reviewing the PRs containing DB migration scripts. Both are marked with a TODO comment. I need help on this, as I have almost no experience with either. Please send any pull request to my `fix-pr-guide` branch, thanks!

----

For convenience, here's the link to the rendered version of the new file - https://github.com/plamut/ggrc-core/blob/fix-pr-guide/doc/pull_request_reviews.md

The existing file - https://github.com/google/ggrc-core/blob/develop/doc/pull_request_reviews.md

zidarsk8 commented on doc/pull_request_reviews.md:
````diff
@@ -1,51 +1,101 @@
-3. Start your dev env. No need if you already have it running
-```
-vagrant up
+    _NOTE: If the merge is not a fast-forward, it could break a few things._
````
This note is wrong, basically even FF can break a few things but the merge must not be a ff, since all PR's are merged with --no-ff flag. So to test that we must make sure we never use FF.

zidarsk8 commented on doc/pull_request_reviews.md:
````diff
@@ -1,51 +1,101 @@
+    ```console
+    git checkout -b temp_branch
+    git fetch <pr_origin>
+    git merge <pr_origin>/<feature_branch_name>
````
this needs a `--no-ff` flag

zidarsk8 commented on doc/pull_request_reviews.md:
````diff
@@ -1,51 +1,101 @@
+
+```ini
+[alias]
+    pr = "!f() { git fetch -fu ${2:-origin} refs/pull/$1/head:pr/$1 && git checkout pr/$1; }; f"
````
can you update this alias with 

    [alias]                                                                     
      pr = "!f() { \                                                            
        git checkout develop; \                                                 
        git branch -D pr-$1; \                                                  
        git fetch upstream develop:pr-$1; \                                     
        git checkout pr-$1; \                                                   
        git fetch upstream pull/$1/head; \                                      
        git merge --no-ff FETCH_HEAD -m \"Automatic merge\"; \                          
        }; f"

zidarsk8 commented on doc/pull_request_reviews.md:
````diff
@@ -1,51 +1,101 @@
+    ```console
+    vagrant up
+
+    # run the following if there were any changes in the provision/ directory
````
if there were any changes in the provision, or requirrements or dev-requiremens, or npm requirements....

zidarsk8 commented on doc/pull_request_reviews.md:
```diff
@@ -1,51 +1,101 @@
+    If you don't yet have the `<pr_origin>` defined, you need to add it
+    ([instructions](https://git-scm.com/book/en/v2/Git-Basics-Working-with-Remotes#Adding-Remote-Repositories)).
+
+3. Start your dev env. No need if you already have it running:
```
This shoud have a small section on docker since we're moving to that.

Although since we have  all of this already written in the main README, we should just say to start the env (docker or vagrant) and reprovision if there's any changes in the provisioning files.

zidarsk8 commented:
Since we're modifying this, I believe this doc should have 2 sections, first would be setup and teardown of the env for PR testing (which is basically what we already have). Second section should be on how and what to actually check on the PR. Including how to marke PRs with labels and such.

Some examples of what else needs to be in this doc:

if the Pull request contains a migration, user must check (also check that the pr has the migration label):
* the migration works from a clean db.
* dowgrade and upgrade works on a clean db (downgrate to previous and upgrade back up)
* that the migrations work from the previous state on the develop branch (do this by, going to an update develop branch, cleaning the db, migrating, switching back to the PR branch and running db_migrate agani)
* That the migration (upgrade and downgrade) work on a full database (for this, users must use a db dump from either grc-test or grc-dev instances)

doing a review of a branch that is a bug fix, a user MUST first reproduce the bug on vanila develop branch before confirming that the bug is fixed with the PR.

Also, user shuld reject a pr if there's no tests written for a bugfix (there are exceptions but a developer will have to give a good reason for it)

plamut commented:
Seeing all the comments, the existing instructions really seem obsolete now, seems like a more thorough rewrite is needed. :)

I will include the suggested changes, though some help would be appreciated (e.g. I don't know that much about Docker). When I'm done, I'll start accepting pull requests into my `fix-pr-guide` branch.

I will

plamut commented:
@zidarsk8 I have incorporated the suggested changes and added a section on how to actually review a pull request. Please proof-read it, thanks!

I also need somebody to write the Docker commands section, and a section describing the extra steps needed when reviewing a PR that contains a DB migration script (I have no experience with these two yet). Please send any such pull request to my `fix-pr-guide` branch, thanks!

plamut commented:
@zidarsk8 How much work do you think it would take to write the Docker commands and the DB migrate sections? (I need someone with more background knowledge to help me with this)

These are the only two missing pieces left, before this PR can be merged IMO.

plamut commented:
@zidarsk8 Please review and finish the Docker section, thanks!

zidarsk8 commented on doc/pull_request_reviews.md:
````diff
@@ -1,51 +1,289 @@
-```
-git checkout develop 
-git branch -D temp_branch
+    _NOTE: The merge must be a fast-forward, since all pull requests are merged
````
The merge must **not** be a fast forward, ergo the --no-ff.

plamut replied:
Meh, lost in refactoring, thanks for noticing!

zidarsk8 commented on doc/pull_request_reviews.md:
```diff
@@ -1,51 +1,289 @@
+  schema. Such pull requests require additional setup and verification steps,
+* `needs work` - a reviewer has concluded that the PR requires additional work
+  before it can be accepted and merged,
+* `next release` - the PR will be merged after the next release and should not
```
After the next release sounds wrong. it can be merged after the current release has branched off of develop.

zidarsk8 commented on doc/pull_request_reviews.md:
````diff
@@ -1,51 +1,289 @@
 ```
 
-6. Go back to your branch and contiue with your work
+_NOTE: The alias assumes that the _remote_ representing the main development
````
you shuld remove the `_` around remote, since all text here is already italic and this makes it look weird.

zidarsk8 commented:
Apart from the few issues above this looks great as for now, and can be merged after those issues are removed.

plamut commented:
@zidarsk8 Resolved all comments. There is still a `TODO` pending for the Docker section, but it's probably not critical because

1. not everybody is using Docker yet,
1. those who do probably already know how to start a test environment.

It would still be nice, though, to add that final missing part in the next few days, but that can be done in a separate PR I guess...
(I can't do it myself as don't yet have a functional Docker on my machine, too old a kernel)

Smotko commented on doc/pull_request_reviews.md:
```diff
@@ -1,51 +1,289 @@
-## Simple helper/guide for reviewing pull requests
+<a name="top"></a>
```
Is this line needed?

plamut replied:
Yes, it's an anchor for all the Back to Top links (found at the end of every major section).

Smotko replied:
I think this is not needed, the top marker can just have a `href="#"` and the browser window will jump to the top.

plamut replied:
I've tried, but a naked `#` sends the user to the very top of the page (including all GitHub menus, headers, etc. that together occupy cca. 1/3 of the screen), instead of to the exact top of the document.

Using an explicit `#top` anchor seems to provide a better experience, although I am not an UX expert of any kind...

Smotko commented on doc/pull_request_reviews.md:
```diff
@@ -1,51 +1,289 @@
+    git stash
+    git checkout develop
+    git fetch <remote_main>
+    git rebase <remote_main>/develop
```
The steps here seem overly complicated. Can we just have something like:

```
git stash # make sure you don't have any local changes
git fetch
git checkout <remote_main>/develop
```

Or maybe even with `git pull`, but I don't really use it.

plamut replied:
I've just used what was there before, but yeah, I like the simplification.

Smotko commented on doc/pull_request_reviews.md:
````diff
@@ -1,51 +1,289 @@
+    ```console
+    git checkout -b temp_branch
+    git fetch <pr_origin>
+    git merge --no-ff <pr_origin>/<feature_branch_name>
````
I'd say we replace this with `git pr 3122` and `git merge <remote_main>/develop` instead of having a tip at the bottom. I want to remove all extra steps so that devs can do this faster/more effective.

plamut replied:
I can't remember exactly what the reason was to list the more verbose version first, but I think it was something about being more explicit about what is going on, helping newcomers to better understand the idea. The "magic" is listed as a convenient shortcut for when these steps become repetitive.

@zidarsk8 Do you maybe remember better what the exact reasoning was in that discussion we had?

zidarsk8 replied:
@plamut I'm gonna say it was done for historic reasons which do not apply anymore.

Smotko commented on doc/pull_request_reviews.md:
```diff
@@ -1,51 +1,289 @@
+    If you don't yet have the `<pr_origin>` defined, you need to add it
+    ([instructions](https://git-scm.com/book/en/v2/Git-Basics-Working-with-Remotes#Adding-Remote-Repositories)).
+
+1. Start your local development environment (Vagrant or Docker). No need if you
```
I think we can assume people know how to set up a dev env here. So instead of this link section I'd prefer if we just mention running `db_migrate` and `deploy_appengine extras/deploy_settings_local.sh` that resets the assets. Also we could mention `launch_gae_ggrc` which simulates appengine better than `launch_ggrc`.

plamut replied:
I personally would not want to accidentally assume too much, even though it might seem "obvious" after awhile for veterans. This step-by-step guide is mainly aimed at newcomers to help them start reviewing faster - they should be able to review a PR by simply following the guide, without having to ask any questions around, and with as few additional cognitive load as possible (when starting, there are way too many things to keep them all in the head, at least that was  an experience I had).

Smotko commented on doc/pull_request_reviews.md:
```diff
@@ -1,51 +1,289 @@
+    git checkout pr-$1; \
+    git fetch upstream pull/$1/head; \
+    git merge --no-ff FETCH_HEAD -m \"Automatic merge\"; \
+  }; f"
```
My git pr alias seems to be a lot simpler:

`pr  = "!f() { git fetch -fu ${2:-<remote_main>} refs/pull/$1/head:pr/$1 && git checkout pr/$1; }; f"`

plamut replied:
Yes, initially that was the alias that was included, but somebody else suggested a "better" alternative that is supposed to be more robust - I think @zidarsk8 or @edofic  (can you two guys comment on that? thanks!)

zidarsk8 replied:
I had some issues with the one @Smotko suggested, but I can't seem to reproduce them. If it works, let's go with the simple version.

Smotko commented on doc/pull_request_reviews.md:
```diff
@@ -1,51 +1,289 @@
+  higher pripority. This label is usually used when the PR has not received
+  enough attention for a considerable time period, or because it is important
+  and attempts to resolve an important/blocking issue,
+* `question` - a reviewer requires additional explanation from the PR author
```
Additionally, `question` can also be used by the person creating the PR if she has a question that needs to be answered. I also think we shouldn't be merging PR before the question label is removed.

plamut replied:
Good point, I will make that explicit, thanks.

plamut commented:
@Smotko Done, reordered the sections so that we have "quickstart" instructions at the top and all the details later on, and replaced the git alias with the former, shorter version.

zidarsk8 commented on doc/pull_request_reviews.md:
```diff
@@ -1,51 +1,314 @@
+[alias]
+    pr = "!f() { \
+        git fetch -fu ${2:-<remote_main>} refs/pull/$1/head:pr/$1 && \
+        git checkout pr/$1; \
```
@Smotko this pr alias does not merge the pr into the dev branch, which means you'll miss the migration chain errors and such. 

If we're using this alias, we should say that this pr-xy branch needs to be merged into the propper upstream branch befero starting the tests.

plamut replied:
So should I put back the old alias? @Smotko ?

Smotko replied:
I don't think the pr command should merge, because what if the PR is not meant to be merged into develop, but into a release branch or some other branch. Also what if the PR has a merge conflict, but you still want to check it out. 

Doing the merge manually is IMHO better, but I'm not the target audience of this doc so you decide what's best for new people.

plamut replied:
@Smotko Valid point, I thus added an extra remark that the PR needs to be tested on the merged branch (which might not always be `develop`), but the alias remained intact, and with no new hardcoded names.
